ompi/mca/coll/acoll/coll_acoll_allgather.c

```diff
@@ -0,0 +1,484 @@
+            int scnt = (send_peer == last_brank) ? last_subgrp_rcnt : bcount;
+            int rcnt = (recv_peer == last_brank) ? last_subgrp_rcnt : bcount;
+
+            tmprecv = (char *) rbuf + (ptrdiff_t) recv_peer * (ptrdiff_t) bcount * rext;
+            tmpsend = (char *) rbuf + (ptrdiff_t) send_peer * (ptrdiff_t) bcount * rext;
+
+            recv_peer <<= log2_sg_size;
+            send_peer <<= log2_sg_size;
+
+            /* Sendreceive */
+            err = ompi_coll_base_sendrecv(tmpsend, scnt, rdtype, sendto,
+                                          MCA_COLL_BASE_TAG_ALLGATHER, tmprecv, rcnt, rdtype,
+                                          recvfrom, MCA_COLL_BASE_TAG_ALLGATHER, comm,
+                                          MPI_STATUS_IGNORE, rank);
+            if (MPI_SUCCESS != err) {
+                return err;
+            }
+        }
+    } else if (!(rank & (sg_size - 1))) {
+        /* Use recursive doubling for power of 2 cases */
+        err = rd_allgather_sub(rbuf, rdtype, comm, bcount, brank, rank, brank, num_sgs, 1, sg_start,
+                               sg_size, rext);
+        if (MPI_SUCCESS != err) {
+            return err;
+        }
+    }
+    /* Now all base ranks have the full data */
+    /* Do broadcast within subgroups from the base ranks for the extra data */
+    if (sg_id == 0) {
+        num_data_blks = 1;
+        data_blk_size[0] = bcount * (num_sgs - 2) + last_subgrp_rcnt;
+        blk_ofst[0] = bcount;
+    } else if (sg_id == num_sgs - 1) {
+        num_data_blks = 1;
+        data_blk_size[0] = bcount * (num_sgs - 1);
+        blk_ofst[0] = 0;
+    } else {
+        num_data_blks = 2;
+        data_blk_size[0] = bcount * sg_id;
+        data_blk_size[1] = bcount * (num_sgs - sg_id - 2) + last_subgrp_rcnt;
+        blk_ofst[0] = 0;
+        blk_ofst[1] = bcount * (sg_id + 1);
+    }
+    reqs = ompi_coll_base_comm_get_reqs(module->base_data, size);
+    if (NULL == reqs) {
+        return OMPI_ERR_OUT_OF_RESOURCE;
+    }
+    nreqs = 0;
+    preq = reqs;
+    /* Loop over data blocks */
+    for (i = 0; i < num_data_blks; i++) {
+        char *buff = (char *) rbuf + (ptrdiff_t) blk_ofst[i] * rext;
+        int sg_dim = opal_hibit(subgrp_size - 1, comm->c_cube_dim);
+        if ((1 << sg_dim) < subgrp_size)
+            sg_dim++;
+        err = sg_bcast_intra(buff, data_blk_size[i], rdtype, rank, sg_dim, size, sg_size, sg_start,
+                             sg_start, comm, module, preq, &nreqs);
+        if (MPI_SUCCESS != err) {
+            ompi_coll_base_free_reqs(reqs, nreqs);
+            return err;
+        }
+    }
+    /* Start and wait on all requests. */
+    if (nreqs > 0) {
+        err = ompi_request_wait_all(nreqs, reqs, MPI_STATUSES_IGNORE);
+        if (MPI_SUCCESS != err) {
+            ompi_coll_base_free_reqs(reqs, nreqs);
+        }
+    }
+
+    /* This barrier is needed to prevent random hangs */
+    err = ompi_coll_base_barrier_intra_tree(comm, module);
```

Comment:
Why the barrier is needed here? This barrier will also add cost to small message allgather.

Reply:
It is removed now.